web/src/copy_and_paste.js

```diff
@@ -330,6 +330,35 @@ export function paste_handler_converter(paste_html) {
     return markdown_text;
 }
 
+function is_safe_url_paste_target($textarea) {
+    const range = $textarea.range();
+
+    if (!range.text) {
+        // No range is selected
+        return false;
+    }
+
+    if (isUrl(range.text.trim())) {
+        // Don't engage our URL paste logic over existing URLs
+        return false;
+    }
+
+    // To avoid doubling up on Markdown link syntax, we want to look
+    // 2 characters back from the original range start for `](`
+    const possible_markdown_link_markers = $textarea.range(range.start - 2, range.start).range();
+    // Reset to the original selection.
+    $textarea.range(range.start, range.end);
+
+    // If the original range start is two or greater, and therefore if
+    // the characters just before the selected range are `](`, assume
+    // we're in or very near existing Markdown link syntax
+    if (range.start >= 2 && possible_markdown_link_markers.text === "](") {
+        return false;
+    }
```

Comment:
I would prefer to just have `if (range.start >= 2) { return true; }` before doing the `possible_markdown_link_markers` calculation -- we should always prefer to do checks for whether a calculation might be invalid before doing the calculation. This logic is (mostly) equivalent, but more likely to be incorrectly refactored in this order.

Also why is `$textarea.range(range.start, range.end);` needed at all -- I don't see what logic above that I would expect to mutate the `$textarea.range()` read above, so the re-select seems suspicious to me...

Reply:
The first point is well taken. I'll make that fix.

The `$textarea.range(range.start, range.end);` is needed to revert to the original selection; it gets mutated on the assignment to `possible_markdown_link_markers`, where we alter the original range to go in search of `](`.

I suppose we could just do some substring work on the value of the `$textarea` using `range.start`, but I don't know that that would be much more readable code. But I can try it and see.